### PR TITLE
 If the user provides an X.Y.Z format version don't contact omnitruck

### PR DESCRIPTION
**Pull request by tas50**
If someone gives us these we do need to lookup a full version string:

12.5
12
:latest

If someone gives us this we don't:

12.13.7

This removes load on our systems, speeds up the run, and allows the run to continue if those servers are down.

Signed-off-by: Tim Smith <tsmith@chef.io>

### CODE REVIEW

**Comment by tas50** [comment]
@zepheiryan @martinisoft @onlyhavecans I think this it the best solution considering some of the constraints we have given user input. Matching the version the user supplied against the version that's installed isn't super simple since allow things like '12' or '12.5' which we have to lookup. If the user provides a full version string though we won't look it up anymore, which will avoid failures if the versions match.

File: providers/default.rb [hunk 1]
```diff
@@ -101,28 +101,29 @@ def mixlib_install
...
-# needs to be expanded to the latest 12.X.
+# the version we WANT TO INSTALL. If the user specifies a version in X.Y.X format
+# we use that without looking it up. If :latest or a non-X.Y.Z format version we
+# look it up with mixlib-install to determine the latest version matching the request
+# @return Mixlib::Versioning::Format::PartialSemVer
 def desired_version
-  if new_resource.download_url_override
-    # probably in an air-gapped environment.
-    version = Mixlib::Versioning.parse(new_resource.version)
-    Chef::Log.debug("download_url_override specified. Using specified version of #{version}")
```
Comment (lamont-granquist) [inline providers/default.rb]: did you lose a use case here?

Comment (tas50) [reply providers/default.rb]: Already fixed that one up. I think that was a bad rebase

**Comment by martinisoft** [comment]
> I think this it the best solution considering some of the constraints we have given user input. Matching the version the user supplied against the version that's installed isn't super simple since allow things like '12' or '12.5' which we have to lookup. If the user provides a full version string though we won't look it up anymore, which will avoid failures if the versions match.

I think it's a good first pass to require full version to verify. A second pass for flexibility (if needed) could be using a caching mechanism to time when it reaches out to check for a newer version similar to how it's done for the apt update resource to prevent constantly checking.

**Comment by tas50** [comment]
@lamont-granquist this should be good for merge / release

**Pull request merged**